demos/jobs/jobs_demo_mosquitto/jobs_demo_mosquitto.c

```diff
@@ -0,0 +1,1264 @@
+        ret = mosquitto_connect( h->m, h->host, h->port, MQTT_KEEP_ALIVE );
+    }
+
+    /* expect the on_connect() callback to update h->connectError */
+    for( i = 0; ( i < MAX_LOOPS ) &&
+         ( ret == MOSQ_ERR_SUCCESS ) &&
+         ( h->connectError == -1 ); i++ )
+    {
+        ret = mosquitto_loop( h->m, MQTT_SHORT_WAIT_TIME, 1 );
+    }
+
+    if( h->connectError > 0 )
+    {
+        warnx( "%s", mosquitto_connack_string( h->connectError ) );
+    }
+    else if( ret != MOSQ_ERR_SUCCESS )
+    {
+        warnx( "connect: %s", mosquitto_strerror( ret ) );
+    }
+
+    return h->connectError == 0 ? true : false;
+}
+
+/*-----------------------------------------------------------*/
+
+static void closeConnection( handle_t * h )
+{
+    assert( h != NULL );
+
+    if( h->m != NULL )
+    {
+        int ret = mosquitto_disconnect( h->m );
+
+        if( ret != MOSQ_ERR_SUCCESS )
+        {
+            warnx( "closeConnection: %s", mosquitto_strerror( ret ) );
+        }
+    }
+}
+
+/*-----------------------------------------------------------*/
+
+static void on_subscribe( struct mosquitto * m,
+                          void * p,
+                          int mid,
+                          int qos_count,
+                          const int * granted_qos )
+{
+    handle_t * h = p;
+
+    assert( h != NULL );
+    assert( granted_qos != NULL );
+    assert( qos_count == 1 );
+
+    /* subscribe() is called with a single topic. */
+    h->subscribeQOS = granted_qos[ 0 ];
+}
+
+/*-----------------------------------------------------------*/
+
+static bool subscribe( handle_t * h,
+                       JobsTopic_t api )
+{
+    int ret;
+    char topic[ JOBS_API_MAX_LENGTH( JOBS_THINGNAME_MAX_LENGTH ) ];
+    JobsStatus_t jobs_ret;
+    size_t i;
+
+    assert( h != NULL );
+    assert( MQTT_QOS <= 2 );
+
+    /* populate the topic buffer with the specified API for use
+     * in a subscription request */
+    jobs_ret = Jobs_GetTopic( topic,
+                              sizeof( topic ),
+                              h->name,
+                              h->nameLength,
+                              api,
+                              NULL );
+    assert( jobs_ret == JobsSuccess );
+    ( void ) jobs_ret;
+
+    /* set to default value */
+    h->subscribeQOS = -1;
+
+    ret = mosquitto_subscribe( h->m, NULL, topic, MQTT_QOS );
+
+    /* expect the on_subscribe() callback to update h->subscribeQOS */
+    for( i = 0; ( i < MAX_LOOPS ) &&
+         ( ret == MOSQ_ERR_SUCCESS ) &&
+         ( h->subscribeQOS == -1 ); i++ )
+    {
+        ret = mosquitto_loop( h->m, MQTT_SHORT_WAIT_TIME, 1 );
+    }
+
+    if( h->subscribeQOS == 0x80 )
+    {
+        warnx( "broker rejected subscription" );
+    }
+    else if( ret != MOSQ_ERR_SUCCESS )
+    {
+        warnx( "subscribe: %s", mosquitto_strerror( ret ) );
+    }
+
+    return ( ( h->subscribeQOS >= 0 ) && ( h->subscribeQOS <= MQTT_QOS ) ) ? true : false;
```

Comment:
So if the broker rejects the subscription, that seems fatal as well. Do you see value in retrying?

Reply:
Re-trying the SUBSCRIBE would make this demo consistent with all other MQTT demos.

Reply:
Please see my comment above.